### PR TITLE
BFD-1789: Leverage Versioning in BFD Deployments

### PR DESCRIPTION
**Pull request by billie-rose**
**JIRA Ticket:**
[BFD-1789](https://jira.cms.gov/browse/BFD-1789)

**User Story or Bug Summary:**

As a BFD release engineer, I want the ability to build and deploy a known tag so that I can run through releases without feeling so silly.

---

### What Does This PR Do?

Enable git tags as valid references when building and deploying services.
Update the Jenkins pipelines (`bfd-build-apps`, `bfd-build-deploy`, `bfd-deploy-apps`, `bfd-build-deployment-process`, `bfd-deploy-pipeline-terraservice` (OOB), and `BFD - Multibranch and Multistage Pipeline` (root `Jenkinsfile`)) to allow tag references similar to branches and PRs.
Fix a bug with ansible where branch tagging was not being applied correctly to AMIs.

This PR was tested against the `2952-test` ephemeral environment using tags based on this branch, e.g. `17.8.9-test[x]`

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [X] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [X] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [X] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [X] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [X] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [X] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [X] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [X] The data dictionary has been updated with any field mapping changes, if any were made.
* [X] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [X] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -135,6 +135,21 @@ try {
...
+			name during PR builds.
+			*/
+			stage('Set Branch Name') {
+				currentStage = env.STAGE_NAME
+				script {
+					if (env.BRANCH_NAME.startsWith('PR')) {
+						gitBranchName = env.CHANGE_BRANCH
+					} else {
+						gitBranchName = env.BRANCH_NAME
+					}
```
Comment (mjburling) [inline Jenkinsfile:149]: Obviously, this is just a move of the existing solution from below to here, but I'm curious about Jenkins behavior in this context. I don't think tag references are enabled for the monolithic job anyway, but we're using this logic elsewhere, too:

* When a _tag reference_ is selected, e.g. `17.8.9-test4`, is this true: `env.BRANCH_NAME == '17.8.9-test4'`?
* Do you know what the behavior of `env.TAG_NAME` is when using a non-tag reference, i.e. a branch?

[Available Environment Variables](https://jenkins-east.cloud.cms.gov/bfd/env-vars.html/) for posterity.

File: ops/jenkins/bfd-build-apps/Jenkinsfile [hunk 2]
```diff
@@ -172,7 +173,7 @@ spec:
                     try {
                         awsAuth.assumeRole()
                         amiIds = scriptForDeploys.buildAppAmis(gitBranchName, gitCommitId, amiIds, appBuildResults)
-                        if (!params.skip_docker_host_ami) {
+                        if (params.force_build_docker_host_ami) {
```
Comment (mjburling) [inline ops/jenkins/bfd-build-apps/Jenkinsfile:176]: 🥇

**Review by mjburling — COMMENTED** [review]
This is some nice, iterative development to something a lot better!

Could we cover this in a more synchronous review today? I have some questions that I can't quite figure out on my lonesome–maybe you've got the answers already?

Top of mind is probably due to my lack of understanding in Jenkins as it relates to the commit hashes represented there. I would think it's straight forward and largely indisputable, but... I could just be bad at this after all. From time to time, I get confused when trying to compare commit hashes between Jenkins (and by extension, AWS resources) and GitHub. Maybe something else is going on under these circumstances (destroyed and recreated tag?) but I assumed that the AMIs generated from e.g. `17.8.9-test4` would match the [tag's commit hash of f3bd858](https://github.com/CMSgov/beneficiary-fhir-data/commit/f3bd858d3877f39345ce591c7c0e3b18a7834a30). Instead... some AMIs sport `2f23cd9d` and the docker-host uses `8a486c3`?

File: Jenkinsfile (hunk 1 above)
Comment (mjburling) [reply Jenkinsfile:149]: As discussed, usage of the more specific `env.TAG_NAME` is **not** something we want to consider at this time. `env.BRANCH_NAME` is equal to the tag reference name.

**Review by mjburling — APPROVED** [review]
🔢

**Review by brandoncruz3 — APPROVED** [review]

**Pull request merged**